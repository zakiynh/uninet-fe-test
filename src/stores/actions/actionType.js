export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_FAILURE = 'LOGIN_FAILURE';
export const LOGOUT = 'LOGOUT';
export const REGISTER_SUCCESS = 'REGISTER_SUCCESS';
export const REGISTER_FAILURE = 'REGISTER_FAILURE';
export const CREATE_POST = 'CREATE_POST';
export const GET_POSTS = 'GET_POSTS';
export const DELETE_POST = 'DELETE_POST';