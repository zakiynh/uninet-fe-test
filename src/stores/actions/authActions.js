import axios from 'axios';
import { LOGIN_SUCCESS, LOGIN_FAILURE, LOGOUT, REGISTER_FAILURE, REGISTER_SUCCESS } from './actionType';
import showError from '../../helpers/swal';

export const login = (credentials) => async (dispatch) => {
  try {
    const response = await axios.post('https://reqres.in/api/login', credentials);
    // Save session in localStorage
    localStorage.setItem('user', JSON.stringify(response.data));
    // Dispatch the login success action if successful
    dispatch({ type: LOGIN_SUCCESS, payload: response.data });
    showError('success', 'Login successful')
  } catch (error) {
    // Dispatch the login failure action if unsuccessful
    dispatch({ type: LOGIN_FAILURE, payload: error.response.data });
    showError('error', error.response.data.error)
  }
};

export const logout = () => (dispatch) => {
  dispatch({ type: LOGOUT });
  localStorage.removeItem('user');
  showError('success', 'Logout successful')
};

export const registerSuccess = () => ({
  type: REGISTER_SUCCESS,
});

export const registerFailure = (error) => ({
  type: REGISTER_FAILURE,
  payload: error,
});

export const register = (userData) => async (dispatch) => {
  try {
    const response = await axios.post('https://reqres.in/api/register', userData);
    dispatch(registerSuccess());
    showError('success', 'Register successful')
  } catch (error) {
    dispatch(registerFailure(error.response.data));
    showError('error', error.response.data.error)
  }
}