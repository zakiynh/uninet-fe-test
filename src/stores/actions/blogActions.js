import { collection, addDoc, getDocs, deleteDoc, doc, updateDoc } from "firebase/firestore";
import { CREATE_POST, GET_POSTS } from "./actionType";
import showError from "../../helpers/swal";
import { firestore } from "../../helpers/firebase";

export const createPost = (post) => async (dispatch) => {
  try {
    const postRef = await addDoc(collection(firestore, "blogPost"), post);
    dispatch({ type: CREATE_POST, payload: postRef });
    showError("success", "Post created successfully");
  } catch (error) {
    showError("error", error.message);
  }
};

export const getPosts = () => async (dispatch) => {
  try {
    const querySnapshot = await getDocs(collection(firestore, "blogPost"));
    const posts = querySnapshot.docs.map((doc) => ({
      ...doc.data(),
      id: doc.id,
    }));
    posts.sort((a, b) => b.createdAt.toDate() - a.createdAt.toDate());
    dispatch({ type: GET_POSTS, payload: posts });
  } catch (error) {
    showError("error", error.message);
  }
};

export const deletePost = (postId) => async (dispatch) => {
  try {
    await deleteDoc(doc(firestore, "blogPost", postId));
    dispatch(getPosts());
    showError("success", "Post deleted successfully");
  } catch (error) {
    console.error("Error deleting post:", error);
    showError("error", error.message);
  }
};

export const updatePost = (postId, updatedPost) => async (dispatch) => {
    try {
      await updateDoc(doc(firestore, 'blogPost', postId), updatedPost);
        showError('success', 'Post updated successfully');
      dispatch(getPosts());
    } catch (error) {
      console.error('Error updating post:', error);
        showError('error', error.message);
    }
  };