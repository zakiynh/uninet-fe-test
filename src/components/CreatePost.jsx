import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { createPost, getPosts } from "../stores/actions/blogActions";
import showError from "../helpers/swal";

const CreatePost = () => {
  const dispatch = useDispatch();
  const [author, setAuthor] = useState("");
  const [title, setTitle] = useState("");
  const [content, setContent] = useState("");

  const handleCreatePost = async () => {
    try {
      if (!author || !title || !content) {
        showError("warning", "Please fill all the fields");
        return;
      }
      const post = {
        author,
        title,
        content,
        createdAt: new Date(),
      };

      await dispatch(createPost(post));
      await dispatch(getPosts());

      setAuthor("");
      setTitle("");
      setContent("");
    } catch (error) {
      console.error("Error creating post:", error);
    }
  };

  return (
    <div className="max-w-2xl mx-auto mt-8 p-4 bg-white rounded-md shadow-md">
    <h2 className="text-2xl font-bold mb-4">Create Blog Post</h2>
    <form className="space-y-4">
        <div className="flex flex-col">
          <label htmlFor="author" className="text-gray-700 font-semibold mb-2">
            Author:
          </label>
          <input
            type="text"
            id="author"
            value={author}
            onChange={(e) => setAuthor(e.target.value)}
            className="border border-gray-300 px-3 py-2 rounded focus:outline-none focus:border-blue-500"
          />
        </div>
        <div className="flex flex-col">
          <label htmlFor="title" className="text-gray-700 font-semibold mb-2">
            Title:
          </label>
          <input
            type="text"
            id="title"
            value={title}
            onChange={(e) => setTitle(e.target.value)}
            className="border border-gray-300 px-3 py-2 rounded focus:outline-none focus:border-blue-500"
          />
        </div>
        <div className="flex flex-col">
          <label htmlFor="content" className="text-gray-700 font-semibold mb-2">
            Content:
          </label>
          <textarea
            id="content"
            value={content}
            onChange={(e) => setContent(e.target.value)}
            className="border border-gray-300 px-3 py-2 rounded focus:outline-none focus:border-blue-500"
            rows="5"
          />
        </div>
        <button
          type="button"
          onClick={handleCreatePost}
          className="bg-blue-500 text-white py-2 px-4 rounded hover:bg-blue-700 focus:outline-none"
        >
          Create Post
        </button>
      </form>
    </div>
  );
};

export default CreatePost;
