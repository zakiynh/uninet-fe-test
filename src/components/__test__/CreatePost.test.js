import React from "react";
import { render, fireEvent, waitForElement } from "@testing-library/react";
import CreatePost from "../CreatePost";
import { useDispatch } from "react-redux";
import { createPost } from "../../stores/actions/blogActions";

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
}));
jest.mock("../../stores/actions/blogActions.js", () => ({
  createPost: jest.fn(),
}));

describe("CreatePost", () => {
  let wrapper;
  let dispatch;
  beforeEach(() => {
    dispatch = jest.fn();
    useDispatch.mockReturnValue(dispatch);
    wrapper = render(<CreatePost />);
  });
  it("should render the component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should dispatch createPost action when form is submitted", async () => {
    const title = wrapper.getByLabelText("title");
    const content = wrapper.getByLabelText("content");
    const submitButton = wrapper.getByText("submit");
    fireEvent.change(title, { target: { value: "Test Title" } });
    fireEvent.change(content, { target: { value: "Test Content" } });
    fireEvent.click(submitButton);
    await waitForElement(() => wrapper.getByText("Post created!"));
    expect(createPost).toHaveBeenCalledWith({
      title: "Test Title",
      content: "Test Content",
    });
    expect(dispatch).toHaveBeenCalledWith({
      type: "CREATE_POST_SUCCESS",
      payload: {
        title: "Test Title",
        content: "Test Content",
      },
    });
  });
});
