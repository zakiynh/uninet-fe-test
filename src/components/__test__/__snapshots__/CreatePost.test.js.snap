// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePost should render the component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="max-w-2xl mx-auto mt-8 p-4 bg-white rounded-md shadow-md"
      >
        <h2
          class="text-2xl font-bold mb-4"
        >
          Create Blog Post
        </h2>
        <form
          class="space-y-4"
        >
          <div
            class="flex flex-col"
          >
            <label
              class="text-gray-700 font-semibold mb-2"
              for="author"
            >
              Author:
            </label>
            <input
              class="border border-gray-300 px-3 py-2 rounded focus:outline-none focus:border-blue-500"
              id="author"
              type="text"
              value=""
            />
          </div>
          <div
            class="flex flex-col"
          >
            <label
              class="text-gray-700 font-semibold mb-2"
              for="title"
            >
              Title:
            </label>
            <input
              class="border border-gray-300 px-3 py-2 rounded focus:outline-none focus:border-blue-500"
              id="title"
              type="text"
              value=""
            />
          </div>
          <div
            class="flex flex-col"
          >
            <label
              class="text-gray-700 font-semibold mb-2"
              for="content"
            >
              Content:
            </label>
            <textarea
              class="border border-gray-300 px-3 py-2 rounded focus:outline-none focus:border-blue-500"
              id="content"
              rows="5"
            />
          </div>
          <button
            class="bg-blue-500 text-white py-2 px-4 rounded hover:bg-blue-700 focus:outline-none"
            type="button"
          >
            Create Post
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="max-w-2xl mx-auto mt-8 p-4 bg-white rounded-md shadow-md"
    >
      <h2
        class="text-2xl font-bold mb-4"
      >
        Create Blog Post
      </h2>
      <form
        class="space-y-4"
      >
        <div
          class="flex flex-col"
        >
          <label
            class="text-gray-700 font-semibold mb-2"
            for="author"
          >
            Author:
          </label>
          <input
            class="border border-gray-300 px-3 py-2 rounded focus:outline-none focus:border-blue-500"
            id="author"
            type="text"
            value=""
          />
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="text-gray-700 font-semibold mb-2"
            for="title"
          >
            Title:
          </label>
          <input
            class="border border-gray-300 px-3 py-2 rounded focus:outline-none focus:border-blue-500"
            id="title"
            type="text"
            value=""
          />
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="text-gray-700 font-semibold mb-2"
            for="content"
          >
            Content:
          </label>
          <textarea
            class="border border-gray-300 px-3 py-2 rounded focus:outline-none focus:border-blue-500"
            id="content"
            rows="5"
          />
        </div>
        <button
          class="bg-blue-500 text-white py-2 px-4 rounded hover:bg-blue-700 focus:outline-none"
          type="button"
        >
          Create Post
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
